graphql-language/src/parser.rs

```diff
@@ -0,0 +1,363 @@
+#![allow(dead_code, unused_variables)]
+use ast::*;
+use lexer::Punctuator::*;
+use lexer::Token::{
+  Eof,
+  Punctuator,
+  Name as TokenName,
+  IntValue as TokenInt,
+  FloatValue as TokenFloat,
+  StringValue as TokenString,
+};
+use lexer::{Lexer, Token};
+use source::Source;
+use std::iter::Peekable;
+
+pub type Res<'a, T> = Result<T, &'a str>;
+pub type DRes<'a, T> = Res<'a, (T, usize)>;
+
+pub struct Parser<'a> {
+  lexer: Peekable<Lexer<'a>>,
+  source: Source<'a>,
+}
+
+impl<'a> Parser<'a> {
+  pub fn new(src: Source<'a>) -> Self {
```

Comment:
I think we should prefer `&mut self`, it feels more idiomatic, especially if we want to move all state into `Parser`. Also it has the advantage of putting all the docs for these methods on a single page. 

Does clippy not give a warning for `&mut Parser`?